Lemma editing in the GUI

felixlinker opened the pull request:
@rsasse and I supervised @4-Pawn's (Alice La Porta) Bachelor thesis, which was about editing lemmas in the GUI. Changes are (more or less) straightforward.

You can now:
- Add lemmas in the GUI
- Delete lemmas from the GUI
- Edit lemmas in the GUI

The hope is this will save the modeler's time by requiring them to restart Tamarin less.

Here's a screenshot of the new GUI:

![image](https://github.com/user-attachments/assets/90f28b50-1bf8-4add-800d-524e3855bdca)

Moreover, we implemented the following:
- When you edit/delete reuse lemmas, subsequent proofs are marked as invalidated, but they are not discarded. There is a button to verify them again.
- There is a button at the top to append all edited/added lemmas to your theory file to persist changes. Some manual merging must be done still, but we thought that appending changes is safer than overwriting. Modelers will certainly not lose any content accidentally. Appending is also safe to the parser, which will stop after reading `end`.

Here's a screenshot of how invalidated proofs look:
![image](https://github.com/user-attachments/assets/42f75a86-08ec-4012-9005-b633a3648f29)

jdreier commented:
Thanks for proposing this nice feature!

Currently the PR does not include documentation, but I think we need to update and extend the manual a bit. In particular the explanation of the GUI in https://github.com/tamarin-prover/tamarin-prover/blob/develop/manual/src/003_example.md needs to be updated (both pictures and the text). The new features themselves could be documented either there or in https://github.com/tamarin-prover/tamarin-prover/blob/develop/manual/src/011_advanced-features.md .

jdreier commented:
Some minor suggestions for improved wording in the GUI:

For lemma editing:

```
- Modifying the lemma in the box above and clicking the submit button will attempt to modify the lemma in the current theory.
- Failures in parsing the lemma or verifying its well-formedness will result in an error, and the lemma will NOT be modified. However, your changes will be kept on this page until you leave this right panel.
- Editing a lemma will NOT modify the file it was loaded from, but clicking on the "append lemmas to file" button adds all modified lemmas as a comment at the end of the file on disk they were loaded from.
- Clicking on the "Download" button will download the modified version of the theory (including the modified lemmas), but not modify the file on disk.
- Modifying a reuse lemma will invalidate all subsequent proofs.
- Modifying a sources lemma is not supported and will result in an error.
```

For deletion:

```                
- Clicking on the button above will delete the lemma from the loaded theory.
- Deleting a lemma will NOT modify the file it was loaded from, but clicking on the "Download" button will download the modified version of the theory (so without the deleted lemmas)
- Deleting a reuse lemma will invalidate all subsequent proofs.
- Deleting a source lemma is not supported and will result in an error
```

For addition:

```
- Adds the lemma in the current position in the theory, but will throw an error if a lemma with the same name exists, the parsing fails, or the lemma isn't well-formed.
- Adding a lemma will NOT modify the loaded source file, but clicking on the "Append lemmas to file" button appends all added lemmas as a comment at the end of the current theory file.
- Clicking on the "Download" button will download the modified version of the theory (including the added lemmas).
```

rsasse commented:
Hey @4-Pawn and @felixlinker , is either of you able to provide some of the documentation that @jdreier is requesting? If so, what timeframe would you envision?

felixlinker commented:
@jdreier I updated the phrasing, screenshots in the example section of the manual, and added new documentation in the "Property Specification" section.

felixlinker commented:
@jdreier @cascremers Rebased onto develop! Can we please merge now? It has been four months since I implemented the requested changes.
